### PR TITLE
feat: ControlExtensions and InputExtensions 

### LINKED ISSUE #280: We need a recipe for customizing keyboard enter key behavior
## What would you like to be added:

We need to have a way to customize the native keyboard enter key behavior so we could for instance implement inter `Textbox` and/or `PasswordBox` navigation. Ideally it would look something like this:

```xmal

<TextBox Text={Binding Username}
               keyboardBehavior.NextControl={Binding ElementName=PasswordField} />

<PasswordBox x:Name="PasswordField"
                        Text={Binding Password}
                        keyboardBehavior.EnterCommand={Binding Login}
                        keyboardBehavior.DismissKeyboardOnEnter="True" />

<Button Content="Login" 
              Command={Binding Login} />

```

So we could have a behavior that could do:
- Redirect focus on another control after the user press the keyboard enter key.
- Execute a command after the user press the keyboard enter key.
- Close the keyboard after the user press the keyboard enter key.

## Why is this needed:

We need this for one of our Uno.WinUI project.

## For which Platform:

- [x] iOS
- [x] Android
- [ ] WebAssembly
- [ ] WebAssembly renders for Xamarin.Forms
- [x] Windows (optional)
- [ ] Build tasks

### LINKED ISSUE #282: Add Command/CommandParameter Extensions for ListViewBase
Add the ability to attach Command/CommandParameter properties to a ListView, this would allow the click of each item in the list to execute the specified ICommand.

### PR DESCRIPTION
**Pull request by Xiaoy312**
GitHub Issue (If applicable): closes #280, closes #282

## PR Type

What kind of change does this PR introduce?

- Feature

## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] Tested code with current [supported SDKs](../README.md#supported)
- [x] Docs have been added/updated which fit [documentation template](https://github.com/nventive/Uno/blob/master/doc/.feature-template.md). (for bug fixes / features)
- [ ] [Unit Tests and/or UI Tests](doc/articles/working-with-the-samples-apps.md) for the changes have been added (for bug fixes / features) (if applicable)
- [ ] [Wasm UI Tests](doc/articles/working-with-the-samples-apps.md#running-the-webassembly-ui-tests-snapshots) are not showing unexpected any differences. Validate PR `Screenshots Compare Test Run` results.
- [x] Contains **NO** breaking changes
- [ ] Updated the [Release Notes](https://github.com/nventive/Uno/tree/master/doc/ReleaseNotes)
- [x] Associated with an issue (GitHub or internal)

### CODE REVIEW

**Comment by Xiaoy312** [comment]
note: this is still in draft, 
as InputExtensions.EnterCommand will be aliased from ControlExtensions.Command
and there is support for ListView item-clicks and navigation-controls navigation to be added

File: doc/helpers/Input-extensions.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
+# InputExtensions Attached Properties
+Provides various attached properties for _input controls_, such as `TextBox` and `PasswordBox`.
+
+## Properties
+Property|Type|Description
+-|-|-
+AutoDismiss|bool|Whether the soft-keyboard will be dismissed when the enter key is pressed.
+AutoFocusNext|bool|Whether the focus will move to the next focusable element when the enter key is pressed.\*
+AutoFocusNextElement|Control|Sets the next control to focus when the enter key is pressed.\*
```
Comment (MatFillion) [inline doc/helpers/Input-extensions.md:9]: I think it would be insightful to write why you offer both the bool and the element.

I'm not sure we need the bool, seems it adds to the contract for low value.

File: doc/helpers/Input-extensions.md [hunk 2]
```diff
@@ -0,0 +1,27 @@
+# InputExtensions Attached Properties
+Provides various attached properties for _input controls_, such as `TextBox` and `PasswordBox`.
+
+## Properties
+Property|Type|Description
+-|-|-
+AutoDismiss|bool|Whether the soft-keyboard will be dismissed when the enter key is pressed.
+AutoFocusNext|bool|Whether the focus will move to the next focusable element when the enter key is pressed.\*
```
Comment (MatFillion) [inline doc/helpers/Input-extensions.md:8]: Should it start with "Is" ?

File: doc/helpers/Input-extensions.md [hunk 3]
```diff
@@ -0,0 +1,27 @@
+# InputExtensions Attached Properties
+Provides various attached properties for _input controls_, such as `TextBox` and `PasswordBox`.
+
+## Properties
+Property|Type|Description
+-|-|-
+AutoDismiss|bool|Whether the soft-keyboard will be dismissed when the enter key is pressed.
```
Comment (MatFillion) [inline doc/helpers/Input-extensions.md:7]: Personal preference: include Keyboard in the name, because that is what is dismissed.

File: doc/helpers/Input-extensions.md (hunk 1 above)
Comment (Xiaoy312) [reply doc/helpers/Input-extensions.md:9]: Maybe I couldve explained better..?:

So you have the option to move the focus to ... on enter:
- the next focusable control: `AutoFocusNext=True`
- a specified control: `AutoFocusNextElement={Binding ElementName=ThatOneOverThere}
The 2nd option is offered as a fallback solution when the 1st option doesnt behave as expected, or as an alternative when you need the focus to jump in a custom way

File: doc/helpers/Input-extensions.md (hunk 2 above)
Comment (francoistanguay) [reply doc/helpers/Input-extensions.md:8]: As it's more of a feature than a property value that can be read, am ok leaving the Is out.

File: doc/helpers/Input-extensions.md (hunk 1 above)
Comment (francoistanguay) [reply doc/helpers/Input-extensions.md:9]: > note: this is still in draft, as InputExtensions.EnterCommand will be aliased from ControlExtensions.Command and there is support for ListView item-clicks and navigation-controls navigation to be added

WE're not creating 2 public contracts, right? Will all be in ControlExtensions.Command, not in InputExtensions.EnterCommand

Comment (Xiaoy312) [reply doc/helpers/Input-extensions.md:9]: @francoistanguay That was still yesterday's code. There is only `ControlExtensions.Command` now.

File: doc/helpers/Input-extensions.md (hunk 3 above)
Comment (Xiaoy312) [reply doc/helpers/Input-extensions.md:7]: > We can actually just call it AutoDismiss, as we're in the context of InputExtensions already 
> \- FT

File: src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+		}
+
+		private static void OnListViewItemClick(object sender, ItemClickEventArgs e)
+		{
+			if (sender is not ListViewBase host) return;
+
+			if (GetCommand(host) is { } command &&
+				command.CanExecute(e.ClickedItem))
+			{
+				command.Execute(e.ClickedItem);
```
Comment (kazo0) [inline src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs:78]: What about a CommandParameter?

File: doc/helpers/Input-extensions.md (hunk 3 above)
Comment (MatFillion) [reply doc/helpers/Input-extensions.md:7]: Input and keyboard are 2 different things :D

File: src/Uno.Toolkit.UI/Behaviors/InputExtensions.cs [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+		#region DependencyProperty: AutoFocusNext
+
+		public static DependencyProperty AutoFocusNextProperty { get; } = DependencyProperty.RegisterAttached(
+			"AutoFocusNext",
+			typeof(bool),
+			typeof(InputExtensions),
+			new PropertyMetadata(default(bool), OnAutoFocusNextChanged));
+
+		public static bool GetAutoFocusNext(DependencyObject obj) => (bool)obj.GetValue(AutoFocusNextProperty);
+		public static void SetAutoFocusNext(DependencyObject obj, bool value) => obj.SetValue(AutoFocusNextProperty, value);
```
Comment (dr1rrb) [inline src/Uno.Toolkit.UI/Behaviors/InputExtensions.cs:45]: Please follow the standard by prefixing you property with `Is` (i.e. `IsAutoFocusNextEanbled`). 

BTW do we really need such property, compared to just relying on the fact that the `IsAutoFocusElement` is set ?
And if you do want to have 2 properties, the default value of this should be `true`.

File: src/Uno.Toolkit.UI/Behaviors/InputExtensions.cs [hunk 6]
```diff
@@ -0,0 +1,150 @@
...
+		#region DependencyProperty: AutoDismiss
+
+		public static DependencyProperty AutoDismissProperty { get; } = DependencyProperty.RegisterAttached(
+			"AutoDismiss",
+			typeof(bool),
+			typeof(InputExtensions),
+			new PropertyMetadata(default(bool), OnAutoDismissChanged));
+
+		public static bool GetAutoDismiss(DependencyObject obj) => (bool)obj.GetValue(AutoDismissProperty);
+		public static void SetAutoDismiss(DependencyObject obj, bool value) => obj.SetValue(AutoDismissProperty, value);
```
Comment (dr1rrb) [inline src/Uno.Toolkit.UI/Behaviors/InputExtensions.cs:33]: Idem, please follow the standard by prefixing your `bool` properties with `Is`.

File: src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+				}
+			}
+			else if (sender is NavigationView nv)
+			{
+				nv.ItemInvoked -= OnNavigationViewItemInvoked;
+				if (GetCommand(sender) is { } command)
+				{
+					nv.ItemInvoked += OnNavigationViewItemInvoked;
+				}
+			}
```
Comment (dr1rrb) [inline src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs:68]: We should log something if no match I think

File: src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+
+		#endregion
+
+		private static void OnCommandChanged(DependencyObject sender, DependencyPropertyChangedEventArgs e)
+		{
+			if (InputExtensions.IsEnterCommandSupportedFor(sender))
+			{
+				// for input controls, this will be implemented in InputExtensions.
+				InputExtensions.OnEnterCommandChanged(sender, e);
+			}
```
Comment (dr1rrb) [inline src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs:47]: That sounds kind of weird for me ...

**Review by dr1rrb — COMMENTED** [review]
Please add XML-doc on all public members. '.md' docs files are nice, but the XML-doc is much easier to use while using those.

File: src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs (hunk 8 above)
Comment (Xiaoy312) [reply src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs:47]: We are supposed to have 2 distinct behaviors: 
1. InputExtensions.EnterCommand can perform 3 things (switch focus, execute command, and/or close soft-keyboard).
2. ControlExtensions.Command execute the command only

Then, FT wanted to "rename" InputExtensions.EnterCommand into ControlExtensions.Command, and so we have this.
ControlExtensions.Command has different 3 implementations: ListView, NavigationView and Input Controls (Text/PasswordBoxes). They are already split in this ControlExtensions, except the input controls one also interacts with other InputExtensions properties, so the implemenation is placed there.

File: src/Uno.Toolkit.UI/Behaviors/InputExtensions.cs (hunk 5 above)
Comment (Xiaoy312) [reply src/Uno.Toolkit.UI/Behaviors/InputExtensions.cs:45]: (The "Is" prefix will be dsicussed there: https://github.com/unoplatform/uno.toolkit.ui/pull/287#discussion_r950437134)

tl;dr: 
```cs
if (AutoFocusNext || AutoFocusNextElement != null)
     Focus(AutoFocusNextElement ?? FindNextFocusable());
```

For having two property here, they are needed because we have 2 DIFFERENT behaviors:
AutoFocusNext and AutoFocusNextElement have different focus target:
- AutoFocusNext is determined by `FocusManager.FindNextFocusableElement`
- AutoFocusNextElement is provided by the value.

File: src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs (hunk 4 above)
Comment (Xiaoy312) [reply src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs:78]: It makes little sense here to have command parameter, since all cases already have a parameter value provided by its context:
- Text/PasswordBox: the Text/Password itself
- ListView: the selected item
- Navigation: the navigated item

But, I will go ahead to add it, for consistency with buttons.

File: doc/helpers/Input-extensions.md (hunk 3 above)
Comment (Xiaoy312) [reply doc/helpers/Input-extensions.md:7]: We are keeping the existing property name _for now_.

File: src/Uno.Toolkit.UI/Behaviors/InputExtensions.cs (hunk 6 above)
Comment (Xiaoy312) [reply src/Uno.Toolkit.UI/Behaviors/InputExtensions.cs:33]: We are keeping the existing property name _for now_.

File: src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs (hunk 7 above)
Comment (Xiaoy312) [reply src/Uno.Toolkit.UI/Behaviors/ControlExtensions.cs:80]: Will do them in a separate PR.
